Add variant=release to Travis CI builds of tests

mloskot opened the pull request:
This aims to confirm there are no issues depending on build configuration (related to #46 and #49).

## References

#51 

-----

BTW, `variant=debug,release` is intentional, it first builds tests with `variant=debug`, then `variant=release`.

stefanseefeld commented:
The `channel` test fails whenever I compile the "channel.cpp" file with -O3. I'm looking into it now...

stefanseefeld commented:
...in the `test_channel_convert()` function, a call to `channel_convert<>` results in wrong values. This could indeed be a manifestation of the same issue as in https://github.com/boostorg/gil/issues/49.

chhenning commented:
@stefanseefeld Can you make a minimal code sample?

stefanseefeld commented:
Yes, and I made an interesting observation:

I narrowed the error down to a call to ` do_test<reference_core<bits32&> >().test_all()`.
As the next step I "unwrapped" that function, replacing it with
```
   do_test<reference_core<bits32&> >().test_channel_invert();
   do_test<reference_core<bits32&> >().test_channel_convert();
   do_test<reference_core<bits32&> >().test_channel_multiply();
   do_test<reference_core<bits32&> >().test_channel_math();
```
which made the error disappear ! Notice, however, that this isn't strictly equivalent. In the original case above the `test_all()` call reuses the `do_test` object for the four test functions, while in my "unwrapped" case I construct new temporary `do_test` objects for each of those functions.
Thus suggests that the problem is that the object's state gets altered in a way to make a subsequent test function call fail.
@chhenning does this ring a bell ? Can you think of any possible reason for that behaviour, which is also specific to this using channel *references*...

stefanseefeld commented:
(You may just drop the above code right into `main()` to reproduce the problem. Don't wrap it in a `try` block, so a failure will cause an "unexpected" exception, allowing you to inspect the core dump...)

stefanseefeld commented:
@chhenning , looking into the `channel.hpp` file, I notice quite a few classes (`packed_channel_reference_base` and all its subclasses) having modifiable state, yet all member functions being declared `const`. What's up with that ? Can you explain the meaning of `const` there ?

mloskot commented:
I guess, this can be confirmed as a bug

chhenning commented:
Ok. I have set up visual studio code on ubuntu and with boost gil master the channel.cpp test runs through. I'm using gcc 5.4.something

This would imply that gil develop different gil core which is weird.

mloskot commented:
@chhenning 
> boost gil master the channel.cpp test runs through. I'm using gcc 5.4.something

What compilation flags exactly?

chhenning commented:
Yeah, I was just testing debug configuration.

chhenning commented:
On gil develop I can see an exception with release and release_with_debug_symbols. But no exception with debug config.

mloskot commented:
@chhenning That seems to confirm the issue.

To summary, it occurs only for optimised builds only.
- Linux+GCC, #51, https://travis-ci.org/boostorg/gil/jobs/355009262
- Linux+clang, 

**Question**: does it occur for optimised builds for both, 32bit and 64bit targets or olny for 64bit, as it interestingly does in case of the checksum bug, #49, https://ci.appveyor.com/project/stefanseefeld/gil/build/1.0.182-develop

stefanseefeld commented:
It would be useful to pay close attention to the compiler warnings. (Disable those about unused local typedefs to reduce the noise.) There are a few from `channel.hpp` about using uninitialized memory. I didn't quite understand that warning, it looked bogus. But then, it may hint at the problem.
(Inserting print statements to print out the `max` and `min` values in the test makes the failure disappear, so I'm pretty sure this has to do with illegally aliased data that the optimization pass may actually elide, or some such. I haven't looked at `clang`s warnings yet.

mloskot commented:
@stefanseefeld @chhenning It may be a good time to start shooting down the warnings.

To avoid chasing our own and each others tails, we rather get somewhat systematic about it.

Let's agree on some plan:

1. Create `fix-warnings` shared branch in the `boostorg/gil`
2. Temporarily disable the failing tests
3. Compile using clang, gcc, msvc (`/W3` level but perhaps `/W4` too)
4. Pick warning, fix it, commit 
5. Repeat 2-3 until we are are warnings-free and the builds are green
6. Re-enable failing tests and see what happens.
7. If this does not help, we run static analysis and repeat the clean-up cycles.

Comments?

chhenning commented:
Here is my minimal reproducible code
```
#include <iostream>
#include <boost/gil/gil_all.hpp>

using namespace boost::gil;
using namespace std;

void error_if(bool condition) {
    if (condition)
        throw std::exception();
}

bits32s c32s_min = channel_traits<bits32s>::min_value();
bits32s c32s_max = channel_traits<bits32s>::max_value();

// For channel values simply initialize the value directly
template <typename ChannelValue>
struct value_core {
    typedef ChannelValue channel_t;
    channel_t _min_v, _max_v;

    value_core() : _min_v(channel_traits<ChannelValue>::min_value()), _max_v(channel_traits<ChannelValue>::max_value()) 
    {}
};

int main(int argc, char *argv[])
{
    value_core<bits32s> a;

    bits32s v_min, v_max;

    v_min = channel_convert<bits32s>(c32s_min);
    v_max = channel_convert<bits32s>(c32s_max);

    //std::cout << v_min << std::endl;
    // std::cout << this->_min_v << std::endl;
    //std::cout << v_max << std::endl;
    // std::cout << this->_max_v << std::endl;

    error_if(v_min != a._min_v || v_max != a._max_v);  
}
```

chhenning commented:
Funny enough, when enabling the cout's. All is good.

chhenning commented:
To expedite the issue I sent an email to boost mailing list.

mloskot commented:
I've read the post on ML, I'd insist we do need better coverage of compilers and configs on CI builds, plus static analysis. Sorry, we can't save energy there, unless we have too much time.

stefanseefeld commented:
I have never argued against better coverage. I'v been arguing against redundancy (building the same variants on multiple CI systems, for example, or building unnecessarily often).

mloskot commented:
I'm sorry if I misunderstood. I'm now on street. I'll be back to this later tonight.

stefanseefeld commented:
@mloskot , following a suggestion from the Boost ML, I tried out this patch:
```
--- a/include/boost/gil/channel_algorithm.hpp
+++ b/include/boost/gil/channel_algorithm.hpp
@@ -329,7 +329,7 @@ template <> struct channel_convert_to_unsigned<bits16s> : public std::unary_func
 
 template <> struct channel_convert_to_unsigned<bits32s> : public std::unary_function<bits32s,bits32> {
     typedef bits32 type;
-    type operator()(bits32s x) const { return static_cast<bits32>(x+(1<<31)); }
+    type operator()(bits32s x) const { return static_cast<bits32>(x)+(1<<31); }
 };
```

which works for me. Can you try that in your branch (i.e., this PR) ?

mloskot commented:
@stefanseefeld I've just updated the PR with the extra commit.

mloskot commented:
I've reverted 9d9fc09a8ffb39667bc27625285d2d8337ba4034 that temporarily disabled the failing checksum test cases and updated this PR to see if the UB fix also fixes the checksum bug #49

mloskot commented:
I've updated the build jobs layout adding multiple compilers: GCC 5, 6, 7 and clang 3.9 X `variant=release,debug`. I defined `variant=release` as allowed to failure builds for now, until we solve the checksum/channel bugs.
